include/ccf/pal/measurement.h

```diff
@@ -0,0 +1,166 @@
+  }
+
+  template <size_t N>
+  inline void from_json(
+    const nlohmann::json& j, AttestationMeasurement<N>& measurement)
+  {
+    if (j.is_string())
+    {
+      measurement = j.get<std::string>();
+    }
+    else
+    {
+      throw JsonParseError(fmt::format(
+        "Attestation measurement should be hex-encoded string: {}", j.dump()));
+    }
+  }
+
+  template <size_t N>
+  inline void fill_json_schema(
+    nlohmann::json& schema, const AttestationMeasurement<N>*)
+  {
+    schema["type"] = "string";
+
+    // According to the spec, "format is an open value, so you can use any
+    // formats, even not those defined by the OpenAPI Specification"
+    // https://swagger.io/docs/specification/data-models/data-types/#format
+    schema["format"] = "hex";
+    schema["pattern"] =
+      fmt::format("^[a-f0-9]{}$", AttestationMeasurement<N>::size() * 2);
+  }
+
+  // SGX
+  static constexpr size_t sgx_attestation_measurement_size = 32;
+  using SgxAttestationMeasurement =
+    AttestationMeasurement<sgx_attestation_measurement_size>;
+
+  inline std::string schema_name(const SgxAttestationMeasurement*)
+  {
+    return "SgxAttestationMeasurement";
+  }
+
+  // SNP
+  static constexpr size_t snp_attestation_measurement_size = 48;
+  using SnpAttestationMeasurement =
+    AttestationMeasurement<snp_attestation_measurement_size>;
+
+  inline std::string schema_name(const SnpAttestationMeasurement*)
+  {
+    return "SnpAttestationMeasurement";
+  }
+
+  // Generic wrapper for code digests on all TEE platforms
+  struct PlatformAttestationMeasurement
+  {
+    std::vector<uint8_t> data;
+
+    PlatformAttestationMeasurement() = default;
+    PlatformAttestationMeasurement(const PlatformAttestationMeasurement&) =
+      default;
+
+    template <size_t N>
+    PlatformAttestationMeasurement(
+      const AttestationMeasurement<N>& measurement) :
+      data(measurement.measurement.begin(), measurement.measurement.end())
+    {}
+
+    PlatformAttestationMeasurement& operator=(
+      const PlatformAttestationMeasurement&) = default;
+
+    std::string hex_str() const
+    {
+      return ds::to_hex(data);
+    }
+
+    operator std::span<const uint8_t>() const
+    {
+      return data;
+    }
+  };
+  DECLARE_JSON_TYPE(PlatformAttestationMeasurement);
+  DECLARE_JSON_REQUIRED_FIELDS(PlatformAttestationMeasurement, data);
+}
+
+namespace kv::serialisers
+{
+  template <size_t N>
+  struct BlitSerialiser<ccf::pal::AttestationMeasurement<N>>
+  {
+    static SerialisedEntry to_serialised(
+      const ccf::pal::AttestationMeasurement<N>& measurement)
+    {
+      auto hex_str = measurement.hex_str();
+      return SerialisedEntry(hex_str.begin(), hex_str.end());
+    }
+
+    static ccf::pal::AttestationMeasurement<N> from_serialised(
+      const SerialisedEntry& data)
+    {
+      ccf::pal::AttestationMeasurement<N> ret;
+      ds::from_hex(std::string(data.data(), data.end()), ret.measurement);
+      return ret;
```

Comment:
Why do we convert to hex here? Is this for compatibility with what we did previously?

Reply:
Yes exactly. This used to be done for `CodeDigest` here: https://github.com/microsoft/CCF/blob/a5854de93bff07d4a830f8725015e012d4f07205/include/ccf/service/code_digest.h#L77-L95

Reply:
Cool, thanks for clarifying. This is a shame, and I _think_ means that our `read_ledger` and similar tooling are obfuscating the measurements here (displaying hex-of-hex), but this is now necessary for backwards compatibility.

Reply:
Good point. Based on https://github.com/microsoft/CCF/blob/main/python/ccf/read_ledger.py#L58, it looks like `read_ledger.py` simply displays the hex string so I don't think the hex-of-hex is displayed anywhere.

Reply:
You're right again - we assume our blitted governance table keys are readable and display them directly as strings, without re-encoding as hex. Briefly validated with sandbox:

```
$ curl https://127.0.0.1:8000/gov/kv/nodes/code_ids -k
{"064a890f04d6bd85e915697023686cacf2cfcd85f9739d0f83054f351fe44cbf":"AllowedToJoin"}
```
```
$ /opt/openenclave/bin/oesign dump -e ./samples/apps/logging/liblogging.enclave.so.signed 
...
=== SGX Enclave Properties:
...
mrenclave=064a890f04d6bd85e915697023686cacf2cfcd85f9739d0f83054f351fe44cbf
```
```
$ read_ledger.py --uncommitted ./workspace/sandbox_0/0.ledger/
...
    table "public:ccf.gov.nodes.code_ids" (1 write):
      064a890f04d6bd85e915697023686cacf2cfcd85f9739d0f83054f351fe44cbf:
        "AllowedToJoin"
...
```